vexination/walker/src/server.rs

```diff
@@ -42,103 +54,166 @@ pub async fn run(
-                            .send()
-                            .await
-                        {
-                            Ok(r) if r.status() == StatusCode::CREATED => {
-                                log::info!(
-                                    "VEX ({}) of size {} stored successfully",
-                                    doc.document.tracking.id,
-                                    &data[..].len()
-                                );
-                            }
-                            Ok(r) => {
-                                log::warn!("(Skipped) {name}: Error storing VEX: {}", r.status());
+    let total = Arc::new(AtomicUsize::default());
+    let duplicates: Arc<std::sync::Mutex<Duplicates>> = Default::default();
+    let errors: Arc<std::sync::Mutex<BTreeMap<DocumentKey, String>>> = Default::default();
+    let warnings: Arc<std::sync::Mutex<BTreeMap<DocumentKey, Vec<CheckError>>>> = Default::default();
+
+    {
+        let client = Arc::new(client.build()?);
+        let total = total.clone();
+        let duplicates = duplicates.clone();
+        let errors = errors.clone();
+        let warnings = warnings.clone();
+
+        let visitor = move |advisory: Result<
+            VerifiedAdvisory<ValidatedAdvisory, &'static str>,
+            VerificationError<ValidationError, ValidatedAdvisory>,
+        >| {
+            (*total).fetch_add(1, Ordering::Release);
+
+            let errors = errors.clone();
+            let warnings = warnings.clone();
+            let client = client.clone();
+            let sink = sink.clone();
+            let provider = provider.clone();
+            async move {
+                let adv = match advisory {
+                    Ok(adv) => {
+                        let sink = sink.clone();
+                        let name = adv
+                            .url
+                            .path_segments()
+                            .and_then(|s| s.last())
+                            .unwrap_or_else(|| adv.url.path());
+
+                        match serde_json::to_string(&adv.csaf.clone()) {
+                            Ok(b) => {
+                                const MAX_RETRIES: usize = 10;
+                                for retry in 0..MAX_RETRIES {
+                                    match client
+                                        .post(sink.clone())
+                                        .header(header::CONTENT_TYPE, "application/json")
+                                        .body(b.clone())
+                                        .inject_token(&provider)
+                                        .await?
+                                        .send()
+                                        .await
+                                    {
+                                        Ok(r) if r.status() == StatusCode::CREATED => {
+                                            log::info!("VEX ({}) stored successfully", &adv.csaf.document.tracking.id);
```

Comment:
Probably should break out of the loop upon success, yeah?

Reply:
Yes!

Reply:
Can we use a `SendVisitor` instead? To better encapsulate the retry behavior. Perhaps incorporate these changes within #1097 ?